ch05construction/09patterns.ipynb

```diff
@@ -2,50 +2,74 @@
-   }
-  },
-  {
-   "cell_type": "markdown",
-   "source": [
-    "### Class Complexity"
-   ],
-   "metadata": {
-    "collapsed": false
-   }
+   ]
   },
   {
    "cell_type": "markdown",
+   "metadata": {},
    "source": [
+    "### Class Complexity\n",
+    "\n",
     "\n",
     "We've seen that using object orientation can produce quite complex class structures, with classes owning each other, instantiating each other,\n",
     "and inheriting from each other.\n",
     "\n",
     "There are lots of different ways to design things, and decisions to make.\n",
     "\n",
-    "> - To reuse functionality: should I use inheritance, or add class variable which is delegated to? (\"is a\" vs \"has a\")\n",
     "> - How much flexibility should I allow in this class's inner workings?\n",
-    "> - Should I split this related functionality into multiple classes or keep it in one?\n"
-   ],
-   "metadata": {
-    "collapsed": false
-   }
+    "> - Should I split this related functionality into multiple classes or keep it in one?\n",
+    "> - To reuse functionality: should I use inheritance, or add class variable which it is delegated to?"
+   ]
   },
   {
    "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Inheritance vs composition\n",
+    "\n",
+    "The last point is known as `is-a` vs `has-a` or inheritance vs composition.\n",
+    "\n",
+    "Both options allow us to define a way for classes to collaborate while also having a single responsibility.\n",
+    "As a rule of thumb, we suggest choosing composition over inheritance unless you have a strong reason.\n",
+    "As we go into below, composition introduces fewer dependencies and assumptions about how your code will be used in the future.\n",
+    "\n",
+    "**Inheritance** *(is-a)*:\n",
+    "\n",
+    "- Used if you want to have the same functionality across multiple instances\n",
+    "- Abstracting commonly used methods or data, so all children can use the generic functionality\n",
+    "    - This has a drawback because we make large assumptions about how the code will be used in the future,\n",
+    "      which is often hard or impossible to do the first time\n",
+    "- If we find a bug in the functionality, this can be fixed in one place\n",
+    "    - This is good because one change can be cascaded\n",
+    "    - This can be bad because every subclass relies on its parent and changing one may need us to change the other\n",
+    "      (known tight coupling)\n",
+    "\n",
+    "\n",
+    "**Composition** *(has-a)*:\n",
+    "\n",
+    "- Create separate classes (component) which carry out the shared functionality.\n",
+    "- Instead of inheriting a method, instantiate the class with the components as class variables,\n",
+    "  when that functionality is required then we call that method on from the component.\n",
+    "- Unlike inheritance, you can design each component's interface so that it knows how to interact with other parts of your code\n",
```

Comment:
Maybe more explicitly? Unless that's not what you meant.
```suggestion
    "- Unlike inheritance, you can design each component's interface independently so that it knows how to interact with other parts of your code\n",
```